File: src/com/blackoutburst/bogel/maths/Vector3i.java
package com.blackoutburst.bogel.maths;

/**
 * <h1>Vector3i</h1>
 * 
 * <p>
 * Create and manager Vector3
 * </p>
 * 
 * @since 0.1
 * @author Blackoutburst
 */
public class Vector3i {

	/**x value*/
	public int x;
	
	/**y value*/
	public int y;
	
	/**z value*/
	public int z;
	
	/**
	 * <p>
	 * Create a new vector (0, 0, 0)
	 * </p>
	 * 
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public Vector3i() {
		this.x = 0;
		this.y = 0;
		this.z = 0;
	}
	
	/**
	 * <p>
	 * Create a new vector (size, size, size)
	 * </p>
	 *
	 * @param size the vector size
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public Vector3i(int size) {
		this.x = size;
		this.y = size;
		this.z = size;
	}
	
	/**
	 * <p>
	 * Create a new vector (x, y, z)
	 * </p>
	 *
	 * @param x the vector x
	 * @param y the vector y
	 * @param z the vector z
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public Vector3i(int x, int y, int z) {
		this.x = x;
		this.y = y;
		this.z = z;
	}
	
	/**
	 * <p>
	 * Set vector values (x, y, z)
	 * </p>
	 *
	 * @param x the vector x
	 * @param y the vector y
	 * @param z the vector z
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public void set(int x, int y, int z) {
		this.x = x;
		this.y = y;
		this.z = z;
	}
	
	/**
	 * <p>
	 * Get vector X value
	 * </p>
	 *
	 * @return int
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public int getX() {
		return (x);
	}

	/**
	 * <p>
	 * Set vector X value
	 * </p>
	 *
	 * @param x the vector x
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public void setX(int x) {
		this.x = x;
	}

	/**
	 * <p>
	 * Get vector Y value
	 * </p>
	 *
	 * @return int
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public int getY() {
		return (y);
	}

	/**
	 * <p>
	 * Set vector Y value
	 * </p>
	 *
	 * @param y the vector y
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public void setY(int y) {
		this.y = y;
	}
	
	/**
	 * <p>
	 * Get vector Z value
	 * </p>
	 *
	 * @return int
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public int getZ() {
		return (z);
	}

	/**
	 * <p>
	 * Set vector Z value
	 * </p>
	 *
	 * @param z the vector z
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public void setZ(int z) {
		this.z = z;
	}
	
	/**
	 * <p>
	 * Normalize the vector
	 * </p>
	 *
	 * @return Vector3i
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public Vector3i normalize() {
		float mag = (float) Math.sqrt(x * x + y * y + z * z);
		if (mag != 0) {
			x /= mag;
			y /= mag;
			z /= mag;
		}
		
		return (this);
	}
	
	
	/**
	 * <p>
	 * Multiply two vector
	 * </p>
	 *
	 * @param v the vector scale
	 * @return Vector3i
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public Vector3i mul(float v) {
		x *= v;
		y *= v;
		z *= v;
		
		return (this);
	}
	
	/**
	 * <p>
	 * 	Return the vector length
	 * </p>
	 *
	 * @return float
	 * @since 0.1
	 * @author Blackoutburst
	 */
	public float length() {
		return ((float) Math.sqrt(x * x + y * y + z * z));
	}
	
	/**
	 * <p>
	 * Create a copy of the Vector
	 * </p>
	 * 
	 * @return Vector3i
	 * @since 0.2
	 * @author Blackoutburst
	 */
	public Vector3i copy() {
		Vector3i newVector = new Vector3i();
		newVector.x = this.x;
		newVector.y = this.y;
		newVector.z = this.z;
		
		return (newVector);
	}

	@Override
	public String toString() {
		return "["+x+", "+y+", "+z+"]";
	}
}

File: server/src/main/kotlin/dev/blackoutburst/server/maths/Vector3i.kt
package dev.blackoutburst.server.maths

import kotlin.math.sqrt

class Vector3i {
    var x: Int
    var y: Int
    var z: Int

    constructor() {
        this.x = 0
        this.y = 0
        this.z = 0
    }

    constructor(size: Int) {
        this.x = size
        this.y = size
        this.z = size
    }

    constructor(x: Int, y: Int, z: Int) {
        this.x = x
        this.y = y
        this.z = z
    }

    fun set(x: Int, y: Int, z: Int) {
        this.x = x
        this.y = y
        this.z = z
    }

    fun distance(pos: Vector3i): Int = (pos.x - this.x) * (pos.x - this.x) + (pos.y - this.y) * (pos.y - this.y) + (pos.z - this.z) * (pos.z - this.z)

    fun normalize(): Vector3i {
        val mag = sqrt((x * x + y * y + z * z).toDouble()).toFloat()
        if (mag != 0f) {
            x = (x / mag).toInt()
            y = (y / mag).toInt()
            z = (z / mag).toInt()
        }

        return (this)
    }

    fun mul(v: Float): Vector3i {
        x = (x * v).toInt()
        y = (y * v).toInt()
        z = (z * v).toInt()

        return (this)
    }

    fun length(): Float {
        return (sqrt((x * x + y * y + z * z).toDouble()).toFloat())
    }

    fun copy(): Vector3i {
        val newVector = Vector3i()
        newVector.x = this.x
        newVector.y = this.y
        newVector.z = this.z

        return (newVector)
    }

    override fun toString(): String {
        return "[$x, $y, $z]"
    }


    fun distanceTo(other: Vector3i): Int {
        val dx = (other.x - this.x)
        val dy = (other.y - this.y)
        val dz = (other.z - this.z)
        return dx * dx + dy * dy + dz * dz
    }

    operator fun plus(other: Vector3i) = Vector3i(x + other.x, y + other.y, z + other.z)
    operator fun minus(other: Vector3i) = Vector3i(x - other.x, y - other.y, z - other.z)
    operator fun div(scalar: Int) = Vector3i(x / scalar, y / scalar, z / scalar)
    operator fun times(scalar: Int) = Vector3i(x * scalar, y * scalar, z * scalar)
}
